src/wire_msg.rs

```diff
@@ -67,19 +68,38 @@ impl WireMsg {
             compile_error!("You need an architecture capable of addressing 32-bit pointers");
         }
         // we know we can convert without loss thanks to our assertions above
-        let data_length = msg_header.data_len() as usize;
+        let header_length = msg_header.user_header_len() as usize;
+        let dst_length = msg_header.user_dst_len() as usize;
+        let payload_length = msg_header.user_payload_len() as usize;
+
+        let mut header_data: Vec<u8> = vec![0; header_length];
+        let mut dst_data: Vec<u8> = vec![0; dst_length];
+        let mut payload_data: Vec<u8> = vec![0; payload_length];
 
-        let mut data: Vec<u8> = vec![0; data_length];
         let msg_flag = msg_header.usr_msg_flag();
 
-        recv.read_exact(&mut data).await?;
+        // fill up our data vecs from the stream
+        // ECHO_SRVC_MSG_FLAG does not have header
+        if header_length == 0 {
+            recv.read_exact(&mut header_data).await?;
+        }
+        // ECHO_SRVC_MSG_FLAG does not have dst
+        if dst_length == 0 {
```

Comment:
Shouldn't this be `!= 0`?

Reply:
Yup. and we clearly need some more tests around this too!